### PR TITLE
Fix InkWell overlayColor resolution ignores selected state

### LINKED ISSUE #159063: InkWell overlay colors aren't applied on MaterialState.selected state
### Steps to reproduce

1. Run the code sample.
2. Hover the InkWell square at the center of the screen.

### Expected results

The overlay color is green (the one for selected + hovered states):

https://github.com/user-attachments/assets/1b8481a0-ac6c-4a45-a6ea-34f9b3089459

### Actual results

The overlay color is yellow (selected state is ignored):

https://github.com/user-attachments/assets/8c3351e0-5ad5-48d4-89e1-ec671b3ab79b

### Code sample

<details open><summary>Code sample</summary>

```dart

import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  late WidgetStatesController _controller;

  @override
  void initState() {
    super.initState();
    _controller =
        WidgetStatesController(<WidgetState>{WidgetState.selected});
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final WidgetStateProperty<Color> overlayColor =
        WidgetStateProperty.resolveWith((Set<WidgetState> states) {
      if (states.contains(WidgetState.selected)) {
        if (states.contains(WidgetState.hovered)) {
          return Colors.green;
        }
      }
      if (states.contains(WidgetState.hovered)) {
        return Colors.red;
      }
      return Colors.transparent;
    });

    return MaterialApp(
      home: Directionality(
        textDirection: TextDirection.ltr,
        child: Material(
          child: Center(
            child: InkWell(
              statesController: _controller,
              overlayColor: overlayColor,
              onTap: () {},
              child: Container(
                decoration: BoxDecoration(border: Border.all()),
                width: 100,
                height: 100,
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Flutter (Channel [user-branch], 3.27.0-1.0.pre.538, on macOS 14.2.1 23C71
    darwin-arm64, locale fr-FR)
```

</details>

### PR DESCRIPTION
**Pull request by bleroux**
## Description

This PR fixes `InkWell` overlay colors resolution.
The `InkWell` overlay color is resolved when the `InkWell` is either focused, hovered , and/or pressed.
This resolution happens at two places:
- first when the highlight is created.
- then on each build, using the inner function named `getHighlightColorForType`.
This second resolution should be aware of other current states (such as selected) as it might impact the color resolution.

For instance, several Material styles have colors resolution define similarly to:
https://github.com/flutter/flutter/blob/dc44547d0d90dfff4fd4c3bdcd525f9d7a70744e/packages/flutter/lib/src/material/date_picker_theme.dart#L982-L1006

## Related Issue

Fixes [InkWell overlay colors aren't applied on MaterialState.selected state](https://github.com/flutter/flutter/issues/159063)
First step for [Date picker overlay colors aren't applied on MaterialState.selected state](https://github.com/flutter/flutter/issues/130586).

## Tests

Adds 3 tests.

### CODE REVIEW

File: packages/flutter/lib/src/material/ink_well.dart [hunk 1]
```diff
@@ -1283,12 +1283,16 @@ class _InkResponseState extends State<_InkResponseStateWidget>
...
 
-    Color getHighlightColorForType(_HighlightType type) {
-      const Set<MaterialState> pressed = <MaterialState>{MaterialState.pressed};
-      const Set<MaterialState> focused = <MaterialState>{MaterialState.focused};
-      const Set<MaterialState> hovered = <MaterialState>{MaterialState.hovered};
+    final ThemeData theme = Theme.of(context);
+    const Set<MaterialState> highlightableStates = <MaterialState>{MaterialState.focused, MaterialState.hovered, MaterialState.pressed};
+    final Set<MaterialState> nonHighlightableStates = statesController.value.difference(highlightableStates);
+    // Each highlightable states will be resolved separatly to get the corresponding color.
+    // For this resolution to be correct, The non-highlightable states should be preserved.
```
Comment (justinmc) [inline packages/flutter/lib/src/material/ink_well.dart:1290]: "The" => "the"

File: packages/flutter/lib/src/material/ink_well.dart [hunk 2]
```diff
@@ -1283,12 +1283,16 @@ class _InkResponseState extends State<_InkResponseStateWidget>
...
     super.build(context); // See AutomaticKeepAliveClientMixin.
 
-    Color getHighlightColorForType(_HighlightType type) {
-      const Set<MaterialState> pressed = <MaterialState>{MaterialState.pressed};
-      const Set<MaterialState> focused = <MaterialState>{MaterialState.focused};
-      const Set<MaterialState> hovered = <MaterialState>{MaterialState.hovered};
+    final ThemeData theme = Theme.of(context);
+    const Set<MaterialState> highlightableStates = <MaterialState>{MaterialState.focused, MaterialState.hovered, MaterialState.pressed};
+    final Set<MaterialState> nonHighlightableStates = statesController.value.difference(highlightableStates);
+    // Each highlightable states will be resolved separatly to get the corresponding color.
```
Comment (justinmc) [inline packages/flutter/lib/src/material/ink_well.dart:1289]: "states" => "state"
"separatly" => "separately"

File: packages/flutter/test/material/ink_well_test.dart [hunk 3]
```diff
@@ -11,6 +11,10 @@ import '../widgets/feedback_tester.dart';
 import '../widgets/semantics_tester.dart';
 
 void main() {
+  RenderObject getInkFeatures(WidgetTester tester) {
```
Comment (justinmc) [inline packages/flutter/test/material/ink_well_test.dart:14]: Good call splitting this out.

File: packages/flutter/lib/src/material/ink_well.dart [hunk 4]
```diff
@@ -1283,12 +1283,16 @@ class _InkResponseState extends State<_InkResponseStateWidget>
...
+    // For this resolution to be correct, The non-highlightable states should be preserved.
+    final Set<MaterialState> pressed = <MaterialState>{...nonHighlightableStates, MaterialState.pressed};
```
Comment (justinmc) [inline packages/flutter/lib/src/material/ink_well.dart:1291]: This is kind of tricky, I hope we don't make this same kind of mistake elsewhere. I guess it's probably rare to speculatively resolve a color like this, though.

**Review by justinmc — APPROVED** [review]
LGTM with a few typo nits in the docs :+1:

**Comment by bleroux** [comment]
@justinmc Thanks for the review, especially for catching these typos.

Let see how it goes with Google testing because this might break some golden (if there are golden capturing hovering, focused or pressed state). For instance, with this fix hovered color for some widgets which have a selected state (such as IconButton) is slightly different (the change is slight because the defaults colors are very close but this could be noticed when comparing screenshots.
For instance, here are screenshots of a icon button creating from `IconButton.filledTonal` constructor when hovered:

| Before | After |
|--------|--------|
| Capture d’écran 2024-11-18 à 11 55 48 | Capture d’écran 2024-11-18 à 11 56 08 |

**Pull request merged**